### PR TITLE
 Change Using to implement the loan pattern

### PR DESCRIPTION
**Pull request by NthPortal**
Followup to #7415 and discussion there.
Based on top of #7457 - that can be reviewed an merged independently, or not.

Fixes scala/bug#11225.
Fixes scala/bug#11129.

### CODE REVIEW

**Comment by NthPortal** [comment]
Round 3, I think. I'm pretty happy with this one.

File: src/library/scala/util/Using.scala
Comment (NthPortal) [inline src/library/scala/util/Using.scala:184]: my thought was to save `Managed` for an `apply` method taking an implicit function, once those arrive, which I think makes syntax a bit neater; at that point, this `apply` method on `Manager` could be deprecated

**Comment by SethTisue** [comment]
is there someone(s) who's looked at past iterations of this who could also review this version?

**Comment by NthPortal** [comment]
@dwijnand or @julienrf perhaps?

File: src/library/scala/util/Using.scala
Comment (Ichoran) [inline src/library/scala/util/Using.scala:116]: I don't think this note is going to be enough for most people to understand what to do.  You really need an example, or you probably shouldn't bother with the note.

File: src/library/scala/util/Using.scala
Comment (Ichoran) [inline src/library/scala/util/Using.scala]: This should be `toThrow = t` (one equals)!

File: src/library/scala/util/Using.scala
Comment (Ichoran) [inline src/library/scala/util/Using.scala]: If you made this return the first throwable, the logic would be simpler below.

File: src/library/scala/util/Using.scala
Comment (Ichoran) [inline src/library/scala/util/Using.scala]: These six lines would turn into two if `safeAddSuppressed` returned its first argument.

File: src/library/scala/util/Using.scala
Comment (Ichoran) [inline src/library/scala/util/Using.scala]: Are you sure it's not clearer to have `[R, A](...)(...)(implicit rb: Releasable[R]): A` so you don't have to deal with the `implicitly` stuff below?

File: test/junit/scala/util/UsingTest.scala
Comment (Ichoran) [inline test/junit/scala/util/UsingTest.scala]: Ugh, please don't indent this far.  I recommend

```
def thing(
  arg: Argument,
  arg2: AnotherArgumentThatIsLong,
  bippy: FooBarBazQuux
): Unit = {
  ...
}
```

**Review by Ichoran — CHANGES_REQUESTED** [review]
I like the usage pattern now!

There are some issues (one probably serious--equality test that is mean to be assignment, most minor) but I think it's a good design.  I am unconvinced that the suppression thing is better than creating a new `CompoundException` that just lists everything in order, but I guess since it's already done it should be fine (might be useful).

File: src/library/scala/util/Using.scala
Comment (NthPortal) [reply src/library/scala/util/Using.scala]: Yikes! good catch!

I am very concerned that there is no test which catches that - evidently I missed a case

File: src/library/scala/util/Using.scala
Comment (NthPortal) [reply src/library/scala/util/Using.scala]: true. I may change it, or I may just wait until #7413 is merged, at which point I can just call `addSuppressed`

File: src/library/scala/util/Using.scala
Comment (NthPortal) [reply src/library/scala/util/Using.scala]: good idea.

originally, `Releasable` was called `Resource`, which made it confusing to pass as a parameter since there is already a parameter named `resource`. That's not a problem now though

File: src/library/scala/util/Using.scala
Comment (som-snytt) [reply src/library/scala/util/Using.scala]: There might have been a compiler warning.

File: test/junit/scala/util/UsingTest.scala
Comment (NthPortal) [reply test/junit/scala/util/UsingTest.scala]: Will do.

I didn't do it originally because IntelliJ tries its hardest not to let you

**Comment by NthPortal** [comment]
@Ichoran the problem with a `CompoundException` is you lose the severity and meaning of the original exception, unless you specifically remember to unwrap it. This is particularly problematic for a `VirtualMachineError`, which *must* be propagated so that other code knows to panic and abort.

File: src/library/scala/util/Using.scala
Comment (NthPortal) [reply src/library/scala/util/Using.scala:116]: what do you think of the example I added?

File: src/library/scala/util/Using.scala
Comment (NthPortal) [reply src/library/scala/util/Using.scala]: Note: new added tests catch it

**Comment by NthPortal** [comment]
by the way, for someone merging: the first commit passed, but de-synced. `/sync` will briefly fix it, but not permanently for some reason

**Comment by NthPortal** [comment]
@lrytz could you review the new docs?

File: src/library/scala/util/Using.scala
Comment (julienrf) [inline src/library/scala/util/Using.scala]: “one of them *is* rethrown”?

**Comment by lrytz** [comment]
@NthPortal docs look good to me

**Comment by julienrf** [comment]
Speaking of the docs, I think it would be better to show an example in object `Using` of how to access from files. The current example shows a hypothetical `resource1` but it would be nice to have a full example of how to read a file. (Just reading at the full docs I’m not sure how to use `Using` to read a file)

File: src/library/scala/util/Using.scala
Comment (NthPortal) [reply src/library/scala/util/Using.scala]: Thanks! good catch

Comment (som-snytt) [reply src/library/scala/util/Using.scala]: E.g., comma after `e.g.`  Unless you've already pushed it. I still don't understand how discarding an exception is OK, but I will try it out and comment later when I am better informed.

File: src/library/scala/util/Using.scala
Comment (NthPortal) [inline src/library/scala/util/Using.scala]:
```suggestion
    * garbage collection to be cleaned up would be unacceptable. For example, an instance of
```

File: src/library/scala/util/Using.scala
Comment (NthPortal) [inline src/library/scala/util/Using.scala]:
```suggestion
    implicit val autoCloseableIsReleasable: Releasable[AutoCloseable] = (resource: AutoCloseable) => resource.close()
```

**Comment by NthPortal** [comment]
@julienrf are the new documentation examples satisfactory?

**Comment by julienrf** [comment]
Yes, thanks!

**Comment by NthPortal** [comment]
@Ichoran could you look over this again when you get the chance? I've fixed the major error and added better examples.

File: src/library/scala/util/Using.scala
Comment (Ichoran) [reply src/library/scala/util/Using.scala:116]: The example makes it clearer

**Review by Ichoran — APPROVED** [review]
Looks great to me!  I'd love to use this.  It's similar to but considerably better than what I have in my personal library!

**Comment by NthPortal** [comment]
I think this is good to merge then

**Comment by dwijnand** [comment]
If Julien, Rex and Nth are happy, I'm happy.

**Comment by NthPortal** [comment]
(Note: scala/scala-dev#592 is still an open question)

**Comment by julienrf** [comment]
(that issue didn't generate a lot of discussion, which probably means that people are happy to have it. And anyway, I think this PR does improve the current API)

**Comment by SethTisue** [comment]
@densh since this was partly inspired by your talk (https://www.youtube.com/watch?v=MV2eJkwarT4), perhaps you'd like to do a review pass?

**Pull request merged**
